services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusUpdateRequest.kt

```diff
@@ -0,0 +1,122 @@
+        return nimi != applicationData.name ||
+            katuosoite != applicationData.postalAddress?.streetAddress?.streetName ||
+            rakennustyo != applicationData.constructionWork ||
+            huoltotyo != applicationData.maintenanceWork ||
+            kiinteistoliittyma != applicationData.propertyConnectivity ||
+            hatatyo != applicationData.emergencyWork ||
+            louhintaa != applicationData.rockExcavation ||
+            tyonKuvaus != applicationData.workDescription ||
+            alkuPvm != applicationData.startTime ||
+            loppuPvm != applicationData.endTime ||
+            tyoalueet != applicationData.areas ||
+            yhteystiedot.hasChanges(applicationEntity.yhteystiedot)
+    }
+}
+
+/**
+ * For updating an existing [Hakemusyhteystieto] (with [id]) or creating a new one (without [id]).
+ */
+data class HakemusyhteystietoRequest(
+    /** Hakemusyhteystieto id */
+    val id: UUID? = null,
+    val tyyppi: CustomerType? = null,
+    val nimi: String? = null,
+    val sahkoposti: String? = null,
+    val puhelin: String? = null,
+    val ytunnus: String? = null,
+    val yhteyshenkilot: List<HakemusyhteyshenkiloRequest>? = null,
+) {
+
+    /**
+     * Returns true if this customer has changes compared to the given [hakemusyhteystietoEntity].
+     */
+    fun hasChanges(hakemusyhteystietoEntity: HakemusyhteystietoEntity): Boolean =
+        tyyppi != hakemusyhteystietoEntity.tyyppi ||
+            nimi != hakemusyhteystietoEntity.nimi ||
+            sahkoposti != hakemusyhteystietoEntity.sahkoposti ||
+            puhelin != hakemusyhteystietoEntity.puhelinnumero ||
+            ytunnus != hakemusyhteystietoEntity.ytunnus ||
+            yhteyshenkilot.hasChanges(hakemusyhteystietoEntity.yhteyshenkilot)
+}
+
+data class HakemusyhteyshenkiloRequest(
+    /** Hankekayttaja id */
+    val id: UUID,
+    val tilaaja: Boolean
+)
+
+fun Map<ApplicationContactType, HakemusyhteystietoRequest>?.hasChanges(
+    hakemusyhteystiedot: Map<ApplicationContactType, HakemusyhteystietoEntity>
+): Boolean {
+    if (this == null) {
+        return hakemusyhteystiedot.isNotEmpty()
+    }
+    if (this.size != hakemusyhteystiedot.size) return true
+    return this.entries.any { (role, yhteystietoRequest) ->
+        hakemusyhteystiedot[role]?.let { entity -> yhteystietoRequest.hasChanges(entity) } ?: true
+    }
+}
+
+fun List<HakemusyhteyshenkiloRequest>?.hasChanges(
+    hakemusyhteyshenkilot: List<HakemusyhteyshenkiloEntity>
+): Boolean {
+    if (this.isNullOrEmpty()) {
+        return hakemusyhteyshenkilot.isNotEmpty()
+    }
+    if (this.size != hakemusyhteyshenkilot.size) return true
+    hakemusyhteyshenkilot
+        .map { it.hankekayttaja.id }
+        .toSet()
+        .let { existingIds ->
+            return this.any { hakemusyhteyshenkiloRequest ->
+                hakemusyhteyshenkiloRequest.id !in existingIds
+            }
+        }
```

Comment:
I think this can be simplified. This method could be tested separately to cover all execution paths.
```suggestion
    if (this == null) {
        return hakemusyhteyshenkilot.isNotEmpty()
    }
    val requestIds = map { it.id }.toSet()
    val existingIds = hakemusyhteyshenkilot.map { it.hankekayttaja.id }.toSet()
    return requestIds != existingIds
```

Reply:
Oh, that doesn't cover the tilaaja field.
```suggestion
    if (this == null) {
        return hakemusyhteyshenkilot.isNotEmpty()
    }
    val requests = this.toSet()
    val existing =
        hakemusyhteyshenkilot.map { HakemusyhteyshenkiloRequest(it.id, it.tilaaja) }.toSet()
    return requests != existing
```

Reply:
Come to think of it, the tilaaja field shouldn't be updatable by the user. We're supposed to set it when the the application is sent to Allu.

Reply:
I implemented your suggestions for simplifications. And removed "tilaaja" field. Thanks!